torch/csrc/autograd/input_buffer.cpp

```diff
@@ -74,25 +87,38 @@ static void accumulate(
     Variable&& var) {
   TORCH_INTERNAL_ASSERT(pos < buffer.size());
   auto& old_var = buffer[pos];
-  // ATen doesn't route sparse additions correctly...
-  // do dense + sparse in-place if possible
-  if (old_var.is_sparse()) {
-    // It is safe to change the Tensor inplace if the Tensor is only used in
-    // this buffer (this could be the gradient passed by the user) and that no
-    // other Tensor is using the same storage.
-    if (!var.is_sparse() && var.is_contiguous() && var.use_count() == 1 &&
-        var.storage().use_count() == 1) {
+  // If we hold the last reference to `old_var` AND its storage we will try to
+  // repurpose it to store the output. (Or, if `old_var` is sparse then `var`
+  // becomes the candidate output Tensor.) We only do this if:
+  //  1) GradMode is disabled since Autograd has special handling for inplace
+  //     mutation which we don't want to trigger.
+  //
+  //  2) We hold the last reference.
+  //     (Both `.use_count` and `.storage().use_count()` are one)
+  //
+  //  3) The candidate tensor is a contiguous, non-overlapping, dense, and
+  //     otherwise stock standard Tensor.
+  //
+  //  4) The candidate is mutable. Currently only ZeroTensors are immutable.
+  //
+  //  5) The other Tensor is not a Tensor subclass (except sparse), since
+  //     it's hard to predict the semantics of arbitrary subclass behavior.
+
+  if (at::GradMode::is_enabled()) {
+    buffer[pos] = old_var + var;
+  } else if (
+      // ATen doesn't route sparse additions correctly...
+      old_var.is_sparse() || old_var.is_sparse_csr()) {
```

Comment:
cc @cpuhrsch what is the right way (and future proof way) to check if a Tensor is any sparse format?

Reply:
@albanD - a future proof catch-all is work in progress. Right now it's best to access the layout and compare to the various sparse formats.

`self.layout() == kSparseBsc`, `self.layout() == kSparseBsr`, `self.layout() == kSparseCsc`, `self.layout() == kSparseCsr`, `self.layout() == kSparse` (see [Layout.h](https://github.com/pytorch/pytorch/blob/master/c10/core/Layout.h#L11-L16)).

cc @amjames for visibility. I created a new entry in the roadmap for this so we can track it.